File: src/runtime.ts
import { spawnSync } from 'node:child_process';
import { existsSync, mkdirSync, readFileSync, readdirSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';
import { historyRows } from './analysis.ts';
import { dataDir, quoteHistoryPath, readJsonl, readWatchlist, utcNow } from './storage.ts';
import { tossctlPath } from './toss.ts';

export type RuntimeSnapshot = {
  app: 'TossQuant'; version: string; branch: string; pid: number; tmux: boolean; mode: string; last_action: string;
  watchlist_count: number; watchlist: string[]; quote_files: number; quote_samples: number; quote_counts: Record<string, number>;
  latest_quotes: Record<string, string | null>; classify_ready: string[]; needs_more: string[]; codex: 'ready' | 'missing'; tossctl: string; updated_at: string;
};

export const VERSION = '0.1.0';

export function runtimeDir(base = 'data'): string {
  const path = join(dataDir(base), 'runtime');
  mkdirSync(path, { recursive: true });
  return path;
}

export function runtimeStatePath(base = 'data'): string {
  return join(runtimeDir(base), 'state.json');
}

export function gitBranch(cwd = process.cwd()): string {
  const result = spawnSync('git', ['branch', '--show-current'], { cwd, encoding: 'utf8' });
  return result.status === 0 && result.stdout.trim() ? result.stdout.trim() : 'unknown';
}

export function commandExists(name: string): boolean {
  const result = spawnSync('sh', ['-lc', `command -v ${name}`], { encoding: 'utf8' });
  return result.status === 0;
}

export function quoteSampleCounts(base = 'data'): Record<string, number> {
  const quoteDir = join(dataDir(base), 'quotes');
  if (!existsSync(quoteDir)) return {};
  const counts: Record<string, number> = {};
  for (const file of readdirSync(quoteDir).filter((item) => item.endsWith('.jsonl')).sort()) {
    const ticker = file.replace(/\.jsonl$/, '').toUpperCase();
    try { counts[ticker] = readJsonl(join(quoteDir, file)).length; } catch { counts[ticker] = 0; }
  }
  return counts;
}

export function statusSummary(base = 'data') {
  const counts = quoteSampleCounts(base);
  const watchlist = readWatchlist(base);
  const ready = Object.entries(counts).filter(([, count]) => count >= 3).map(([ticker]) => ticker).sort();
  const needs_more = [...new Set([...watchlist, ...Object.keys(counts)])].filter((ticker) => (counts[ticker] ?? 0) < 3).sort();
  return { watchlist, counts, ready, needs_more };
}

export function latestQuoteTimestamp(ticker: string, base = 'data'): string | null {
  try {
    const rows = historyRows(readJsonl(quoteHistoryPath(ticker, base)));
    return rows.at(-1)?.fetched_at || null;
  } catch { return null; }
}

export function buildRuntimeSnapshot({ mode = 'quant', lastAction = 'ready', base = 'data', cwd = process.cwd() } = {}): RuntimeSnapshot {
  const summary = statusSummary(base);
  const latest: Record<string, string | null> = {};
  for (const ticker of Object.keys(summary.counts)) latest[ticker] = latestQuoteTimestamp(ticker, base);
  return {
    app: 'TossQuant', version: VERSION, branch: gitBranch(cwd), pid: process.pid, tmux: Boolean(process.env.TMUX),
    mode, last_action: lastAction, watchlist_count: summary.watchlist.length, watchlist: summary.watchlist,
    quote_files: Object.keys(summary.counts).length, quote_samples: Object.values(summary.counts).reduce((a, b) => a + b, 0),
    quote_counts: summary.counts, latest_quotes: latest, classify_ready: summary.ready, needs_more: summary.needs_more,
    codex: commandExists('codex') ? 'ready' : 'missing', tossctl: tossctlPath(), updated_at: utcNow(),
  };
}

export function writeRuntimeSnapshot(snapshot: RuntimeSnapshot, base = 'data'): string {
  const path = runtimeStatePath(base);
  writeFileSync(path, `${JSON.stringify(snapshot, null, 2)}\n`, 'utf8');
  return path;
}

export function readRuntimeSnapshot(base = 'data'): RuntimeSnapshot | null {
  const path = runtimeStatePath(base);
  if (!existsSync(path)) return null;
  try { return JSON.parse(readFileSync(path, 'utf8')) as RuntimeSnapshot; } catch { return null; }
}

export function recordRuntime(opts: { mode?: string; lastAction?: string; base?: string; cwd?: string } = {}): RuntimeSnapshot {
  const snapshot = buildRuntimeSnapshot(opts);
  writeRuntimeSnapshot(snapshot, opts.base ?? 'data');
  return snapshot;
}

export function renderRuntimeLine(snapshot: RuntimeSnapshot): string {
  return `[TossQuant ${snapshot.version}] ${snapshot.branch} | mode:${snapshot.mode} | watchlist:${snapshot.watchlist_count} | quotes:${snapshot.quote_files}/${snapshot.quote_samples} samples | classify-ready:${snapshot.classify_ready.length} | codex:${snapshot.codex}`;
}

File: quant_cli_lab/runtime.py
from __future__ import annotations

import json
import os
import shutil
import subprocess
from pathlib import Path
from typing import Any

from . import __version__
from .analysis import history_rows
from .storage import data_dir, quote_history_path, read_jsonl, read_watchlist, utc_now
from . import toss

APP_NAME = "TossQuant"


def runtime_dir(base: str | Path | None = "data") -> Path:
    path = data_dir(base) / "runtime"
    path.mkdir(parents=True, exist_ok=True)
    return path


def runtime_state_path(base: str | Path | None = "data") -> Path:
    return runtime_dir(base) / "state.json"


def _git_branch(cwd: str | Path | None = None) -> str:
    try:
        completed = subprocess.run(
            ["git", "branch", "--show-current"],
            cwd=str(cwd or Path.cwd()),
            text=True,
            capture_output=True,
            check=False,
        )
    except OSError:
        return "unknown"
    branch = (completed.stdout or "").strip()
    return branch or "unknown"


def quote_sample_counts(base: str | Path | None = "data") -> dict[str, int]:
    root = data_dir(base)
    quote_dir = root / "quotes"
    if not quote_dir.exists():
        return {}
    counts: dict[str, int] = {}
    for path in sorted(quote_dir.glob("*.jsonl")):
        try:
            counts[path.stem.upper()] = len(read_jsonl(path))
        except Exception:
            counts[path.stem.upper()] = 0
    return counts


def status_summary(base: str | Path | None = "data") -> dict[str, Any]:
    counts = quote_sample_counts(base)
    watchlist = read_watchlist(base)
    ready = sorted([ticker for ticker, count in counts.items() if count >= 3])
    needs_more = sorted([ticker for ticker in set(watchlist) | set(counts) if counts.get(ticker, 0) < 3])
    return {"watchlist": watchlist, "counts": counts, "ready": ready, "needs_more": needs_more}


def latest_quote_timestamp(ticker: str, base: str | Path | None = "data") -> str | None:
    try:
        rows = history_rows(read_jsonl(quote_history_path(ticker, base)))
    except Exception:
        return None
    if not rows:
        return None
    value = rows[-1].get("fetched_at")
    return str(value) if value is not None else None


def build_runtime_snapshot(
    *,
    mode: str = "quant",
    last_action: str = "ready",
    base: str | Path | None = "data",
    cwd: str | Path | None = None,
) -> dict[str, Any]:
    summary = status_summary(base)
    counts = summary["counts"]
    latest = {ticker: latest_quote_timestamp(ticker, base) for ticker in counts}
    return {
        "app": APP_NAME,
        "version": __version__,
        "branch": _git_branch(cwd),
        "pid": os.getpid(),
        "tmux": bool(os.environ.get("TMUX")),
        "mode": mode,
        "last_action": last_action,
        "watchlist_count": len(summary["watchlist"]),
        "watchlist": summary["watchlist"],
        "quote_files": len(counts),
        "quote_samples": sum(counts.values()),
        "quote_counts": counts,
        "latest_quotes": latest,
        "classify_ready": summary["ready"],
        "needs_more": summary["needs_more"],
        "codex": "ready" if shutil.which("codex") else "missing",
        "tossctl": toss.tossctl_path(),
        "updated_at": utc_now(),
    }


def write_runtime_snapshot(snapshot: dict[str, Any], base: str | Path | None = "data") -> Path:
    path = runtime_state_path(base)
    path.write_text(json.dumps(snapshot, ensure_ascii=False, indent=2, sort_keys=True) + "\n", encoding="utf-8")
    return path


def read_runtime_snapshot(base: str | Path | None = "data") -> dict[str, Any] | None:
    path = runtime_state_path(base)
    if not path.exists():
        return None
    try:
        payload = json.loads(path.read_text(encoding="utf-8"))
    except json.JSONDecodeError:
        return None
    return payload if isinstance(payload, dict) else None


def record_runtime(
    *,
    mode: str = "quant",
    last_action: str = "ready",
    base: str | Path | None = "data",
    cwd: str | Path | None = None,
) -> dict[str, Any]:
    snapshot = build_runtime_snapshot(mode=mode, last_action=last_action, base=base, cwd=cwd)
    write_runtime_snapshot(snapshot, base)
    return snapshot


def render_runtime_line(snapshot: dict[str, Any]) -> str:
    branch = snapshot.get("branch") or "unknown"
    mode = snapshot.get("mode") or "quant"
    watchlist = snapshot.get("watchlist_count", 0)
    quote_files = snapshot.get("quote_files", 0)
    quote_samples = snapshot.get("quote_samples", 0)
    ready = len(snapshot.get("classify_ready") or [])
    codex = snapshot.get("codex") or "missing"
    last_action = snapshot.get("last_action") or "ready"
    updated_at = snapshot.get("updated_at") or "unknown"
    return (
        f"[TossQuant] {branch} | mode:{mode} | watchlist:{watchlist} | "
        f"quotes:{quote_files}/{quote_samples} samples | classify-ready:{ready} | "
        f"codex:{codex} | last:{last_action} | updated:{updated_at}"
    )


def current_runtime_line(
    *,
    mode: str = "quant",
    last_action: str = "ready",
    base: str | Path | None = "data",
    cwd: str | Path | None = None,
) -> str:
    return render_runtime_line(record_runtime(mode=mode, last_action=last_action, base=base, cwd=cwd))
